### PR TITLE
Add tree shaking support for woocommerce components 

### LINKED ISSUE #7024: Add tree shaking support to `woocommerce/components`
The current `woocommerce/components` package is quite big, which makes it more difficult for other plugins to use it if they only need a couple components.

We should add tree shaking support to the package, so it will only bundle in the components used.

This will partly address https://github.com/woocommerce/woocommerce-admin/issues/3289#issue-526709151

### Addition context

p90Yrv-2hM-p2 and https://webpack.js.org/guides/tree-shaking/#root

### PR DESCRIPTION
**Pull request by louwie17**
Fixes #7024 

This adds tree shaking support to the `@woocommerce/components` package, allowing you to only bundle in the used components.
The biggest changes are the `sideEffects` addition in package.json, and the update in `extensions/build.js` to also build the style for each individual component if used outside of WooCommerce.

For the rest none of the components really looked like they had `sideEffects` , and I only updated a couple components that where in the same file.

No changelog as it is in the package.

### Detailed test instructions:

The easiest way of testing this is by making use of the extension examples, just have to update the webpack config for it a bit:
Update the `extensions/examples.config.js` by setting the mode to `production` (enabled tree shaking) adding components to the `excludedExternals` list for the extraction plugin:
```
new WooCommerceDependencyExtractionWebpackPlugin( {
	excludedExternals: [ 'components' ],
} )
```
And I also had to set `loose` to false for the babel-loader (I had some warnings show up), and disabled minification:
```
optimization: {
  minimize: false,
},
```
- Then build one of the examples `npm run example -- --ext=add-task` adding a `woocommerce/components` component to it somewhere. (The Spinner for example).
- Then check the `dist/index.js` file to see if only the imported component and it's dependencies are added, I used a grep command:
`grep -r 'CONCATENATED MODULE: ./packages/components' docs/examples/extensions/add-task/dist`
- Enable the plugin, and see if the component renders correctly (styling will work automatically given we are within WooCommerce).
- Now remove this line to remove the main styling: https://github.com/woocommerce/woocommerce-admin/blob/main/src/Loader.php#L420-L425
- Add the specific component styling to the example `index.scss`  (you might have to enqueue it as well, if it wasn't there before)
Going with the spinner example:
`@import '../../../../../packages/components/build-style/spinner/style.css';`
- Check if the component you added is styled correctly now.

### CODE REVIEW

File: bin/packages/build.js [hunk 1]
```diff
@@ -115,7 +115,7 @@ function buildJsFile( file, silent ) {
  */
 async function buildPackageScss( packagePath ) {
 	const srcDir = path.resolve( packagePath, SRC_DIR );
-	const scssFiles = glob.sync( `${ srcDir }/*.scss` );
+	const scssFiles = glob.sync( `${ srcDir }/**/*.scss` );
```
Comment (louwie17) [inline bin/packages/build.js:118]: This way we generate the css for the individual package components as well.

File: packages/components/package.json [hunk 2]
```diff
@@ -20,6 +20,10 @@
 	"main": "build/index.js",
 	"module": "build-module/index.js",
 	"react-native": "src/index",
+	"sideEffects": [
+		"build-style/**",
+		"src/**/*.scss"
+	],
```
Comment (louwie17) [inline packages/components/package.json:26]: I stole this from `@wordpress/components`, but it seemed to match our use case pretty well.

File: packages/dependency-extraction-webpack-plugin/src/index.js [hunk 3]
```diff
@@ -17,14 +17,18 @@ function camelCaseDash( string ) {
 	return string.replace( /-([a-z])/g, ( _, letter ) => letter.toUpperCase() );
```
Comment (louwie17) [inline packages/dependency-extraction-webpack-plugin/src/index.js:17]: These changes allowed me to more easily exclude a package if needed by doing this:
```
new WooCommerceDependencyExtractionWebpackPlugin( {
	excludedExternals: [ 'components' ],
} ),
```

File: webpack.config.js [hunk 4]
```diff
@@ -119,7 +119,7 @@ const webpackConfig = {
 			},
 			{
 				test: /\.s?css$/,
-				exclude: /storybook\/wordpress/,
+				exclude: [ /storybook\/wordpress/, /build-style\/*\/*.css/ ],
```
Comment (louwie17) [inline webpack.config.js:122]: I added this to not include all the individual component styles as well. Not sure how much of a difference it makes (or if I did it correctly).

File: packages/components/CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,7 @@
+# Unreleased
+
+-   Add tree shaking support to this package. #7034
+
```
Comment (louwie17) [inline packages/components/CHANGELOG.md:4]: If we go ahead with this we probably want to do a major release (7.0.0)

**Comment by louwie17** [comment]
@samueljseay & @psealock I added both of you, given you guy's might be most interested in these changes, let me know what you think?

File: packages/components/src/section/header.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
+/**
+ * External dependencies
+ */
+import { useContext } from 'react';
```
Comment (samueljseay) [inline packages/components/src/section/header.js:4]: I think we should be importing this from `@wordpress/element`

File: packages/dependency-extraction-webpack-plugin/src/index.js [hunk 7]
```diff
@@ -17,14 +17,18 @@ function camelCaseDash( string ) {
...
 }
 
-const wooRequestToExternal = ( request ) => {
+const wooRequestToExternal = ( request, excludedExternals ) => {
```
Comment (psealock) [inline packages/dependency-extraction-webpack-plugin/src/index.js:20]: `excludedExternals` is a new option value, so will need to be reflected in the README

https://github.com/woocommerce/woocommerce-admin/tree/1c16db605af1436647d990230bd146e74f582d88/packages/dependency-extraction-webpack-plugin

**Review by psealock — COMMENTED** [review]
I decided to test this with a brand new plugin using `npm run create-wc-extension`. All of the `@woocommerce/components` get bundled unfortunately, but I may not be doing it right. Here are my steps, but I excluded changes to `babel-loader` because I didn't quite understand what was needed there.

1. `npm run create-wc-extension` to get a new extension
2. `npm link` the dependency extraction webpack plugin so that the new extension can get the version in this branch
3. In my extension, `npm install`
4. `npm install --save @woocommerce/components`
4. `npm link @woocommerce/dependency-extraction-webpack-plugin`
5. `npm run build` Build a production version

My extension's js looks like:

```js
import { Count } from '@woocommerce/components';

console.log( Count );
```
Then I check the build file, `build/index.js` and its pretty clear all the components got included. Can you advise what steps might be missing?

File: packages/dependency-extraction-webpack-plugin/src/index.js (hunk 7 above)
Comment (psealock) [reply packages/dependency-extraction-webpack-plugin/src/index.js:20]: Now that I think about it, this should really be upstream in Gutenberg https://github.com/WordPress/gutenberg/tree/trunk/packages/dependency-extraction-webpack-plugin if you want to make a PR there.

**Comment by louwie17** [comment]
> Then I check the build file, build/index.js and its pretty clear all the components got included. Can you advise what steps might be missing?

I just followed your steps, with the only additional step being `npm link @woocommerce/components` and it seem to tree shake correctly. Maybe you forgot the `npm link @woocommerce/components` step?
The most important addition is the `sideEffects` in package.json, so just see if that is there in `node_modules/@woocommerce/components/package.json`, if not, it didn't link correctly.

File: packages/components/src/section/header.js (hunk 6 above)
Comment (louwie17) [reply packages/components/src/section/header.js]: Updated here: https://github.com/woocommerce/woocommerce-admin/pull/7034/commits/e44d0962157186804e9e720ec535e29920b25235

File: packages/dependency-extraction-webpack-plugin/src/index.js (hunk 7 above)
Comment (louwie17) [reply packages/dependency-extraction-webpack-plugin/src/index.js:20]: Just to clarify: We should keep the changes in this PR (with README updates), but also support the `excludedExternals` param in the original plugin within Gutenberg?

Comment (psealock) [reply packages/dependency-extraction-webpack-plugin/src/index.js:20]: Yup, thats right

**Review by psealock — APPROVED** [review]
>Maybe you forgot the npm link @woocommerce/components step?

Yup, that was the one. Thanks @louwie17 this is testing well for me and is a big improvement for `@woocommerce/components`

File: packages/dependency-extraction-webpack-plugin/src/index.js (hunk 7 above)
Comment (louwie17) [reply packages/dependency-extraction-webpack-plugin/src/index.js:20]: I created a PR, I will just update this PR to follow the format I went with here: https://github.com/WordPress/gutenberg/pull/32220

**Pull request merged**